### PR TITLE
Feature/preferences

### LINKED ISSUE #1038: MacOS Preferences: PreferencesController
This is a project list for the PreferencesController part of the [MacOS Preferences Project](https://github.com/Automattic/simplenote-macos/issues/1037). 

To Dos:

- [ ] Account settings (email, logout, delete account)
- [ ] View Settings (Note Sort Order, Line Length, Condensed Note List, Alphabetical Tags)
- [ ] Theme
- [ ] Text Size
- [ ] Privacy

### PR DESCRIPTION
**Pull request by charliescheer**
NOTE: this is a continuation of #1042  I merged that a bit early.  To see the earlier review/conversation check that PR

### Fix
This is the first PR in the process of adding the Preferences window to Simplenote Mac (over all issue #1037, fixes #1038 )

In this PR I have added all of the UI for the preferences view and linked all of the controls into the window.

Note: one thing that is not done here is the button color for the Delete Account  button.  Still working on getting that going.

### Test
1. Test opening the preferences window by going to Simplenote > Preferences, hitting Command + , and clicking on the Settings button in the tag list
2. Make sure that you can only open one preferences view at a time (shouldn't be possible to have multiple windows of preferences appear)
3. Check and see if your account email appears in the top
4. Test log out button
5. log in with a testing account, test delete account button
6. change the sort order of your notes and make sure the note list updates
7. change the line length setting from full to narrow and back
8. select the condensed note list and deselect it
9. select the sort tags alphabetically and deselect it
10. try the different theme settings
11. use the font slider to change the font size
12. test to make sure these setting changes persist when you close the preferences window and when you restart the app
13. click on the "About Analytics and Privacy" link to be taken to the a8c privacy page

### Release

> `RELEASE-NOTES.txt` was updated in a4948377 with:
> 
> > Added preferences menu to centralize app options and customizations #1042

### CODE REVIEW

File: Simplenote/FontSettings.swift [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+    static func valueIsValidFontSize(_ value: CGFloat) -> Bool {
+        value.truncatingRemainder(dividingBy: step) == .zero
+    }
+
+    static func nearestValidFontSize(from size: CGFloat) -> CGFloat {
+        if valueIsValidFontSize(size) {
+            return size
+        }
+        return 5 * CGFloat((size / 5.0).rounded())
```
Comment (jleandroperez) [inline Simplenote/FontSettings.swift]: This could probably use `step` instead

File: Simplenote/PreferencesViewController.swift [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+            return
+        }
+
+        let isCondensedOn = item.state == .on
+        Options.shared.notesListCondensed = isCondensedOn
+        SPTracker.trackSettingsListCondensedEnabled(isCondensedOn)
+    }
+
+    @IBAction private func sortTagsAlphabeticallyPressed(_ sender: Any) {
+        Options.shared.alphabeticallySortTags = !Options.shared.alphabeticallySortTags
```
Comment (jleandroperez) [inline Simplenote/PreferencesViewController.swift]:
```
Options.shared.alphabeticallySortTags.toggle()
```

Ain't that cool?

**Review by jleandroperez — COMMENTED** [review]
Hey @charliescheer !!! Code wise it's looking good!!

- We should drop the NSMenu items: Delete, Logout, all of the other settings
- Do we need to confirm on Logout? (cc @SylvesterWilmott )
- Clicking the `Logout` button should probably not close the Preferences UI
- Delete account should probably present the confirmation as a `sheet` (rather than a floating alert)

Last but not least, the `Settings` row, in the Tags list, feels super off:

1. Clicking on it doesn't update the Notes List (and you loose track of what's, exactly, the current filter)
2. Closing the `Preferences` Window leaves `Settings` highlighted

@SylvesterWilmott do you think we could remove `Settings` from the tags list?

Thank yooouuu!!

**Comment by SylvesterWilmott** [comment]
This looks great!

After testing the build, I think we should make a few tweaks to the UX:

1. I'm also not too keen on the "Settings" entry point in the tags list. I know it's there for consistency but it's odd to open a window from clicking an item in this panel. I think possibly keeping the "Preferences" menu item is enough. I'm curious to hear other opinions on it before taking any action though.
2. The title in the window title bar is "Settings" but the menu item label is "Preferences". I think we should align on using "Preferences" despite other platforms using "Settings". Mac seems to prefer "Preferences" when referring to settings. Also curious to hear opinions on this one too.

CC: @pachlava 

>Do we need to confirm on Logout?

Logging out immediately after pressing the button does seem a little abrupt but it does match the Log Out action on all other platforms so I'd say it's okay.

>Clicking the Logout button should probably not close the Preferences UI

I think closing the Preferences UI is okay. However if you try to access Preferences whilst viewing the Sign Up/Log In window we should probably hide the "Account" section or potentially disable the "Preferences" menu item.

>Delete account should probably present the confirmation as a sheet (rather than a floating alert)

Agree.

**Comment by pachlava** [comment]
Thanks for asking @SylvesterWilmott ! I've built this branch locally to check, and I do agree that accessing preferences via sidebar feels a bit unusual in the context of apps where `Preferences` is a separate window. I can add the following point supporting its removal from there:

- Since arrow-key navigation is well implemented in the native app, I suspect some ppl might be using it to navigate between Tags / All Notes / Trash. Now, having `Settings` between tags and Trash/All Notes will interrupt this flow. 

But, in the case if there will be a decision to keep it there: I've noticed that once `Settings` is already selected in the sidebar (and `Preferences` are not open), clicking the button once again will not open `Preferences`.

Other notes:
- For some reason, I expected `Esc` key to close `Preferences`. It's not super-widespread in native macOS apps, but one example is Xcode, and I think it's convenient.
- The description label for font size slider in preferences is called `Theme`

cc @charliescheer

**Comment by pachlava** [comment]
One nitpick: I think the button should say `Log Out` / `Log out` instead of `Logout` - this is so on other platforms.

**Comment by SylvesterWilmott** [comment]
Thanks @pachlava. Just going to list the changes here:

1. Remove 'Settings' item from tags list.
2. Update `Theme` label near the text size slider to `Text Size`.
3. Include `Esc` shortcut to close preferences.
4. Update `Settings` title  on preferences window to `Preferences`.
5. Update `Logout` button label to `Log Out`.
6. If the app is in a logged out state, then hide the 'Account' section in preferences.
7. Use a sheet for 'delete account' confirmation instead of dialog.

**Comment by charliescheer** [comment]
Hey @SylvesterWilmott thanks for taking a look.  Sorry for the delay in getting back to this.  Haven't had a chance to find the time since moving over to D1.  I have gone ahead and done all of the following.  You can take another look when you have a chance.

- [x] Remove 'Settings' item from tags list.
- [x] Update Theme label near the text size slider to Text Size.
- [x] Include Esc shortcut to close preferences.
- [x] Update Settings title on preferences window to Preferences.
- [x] Update Logout button label to Log Out.
- [x] If the app is in a logged out state, then hide the 'Account' section in preferences.
- [x] Use a sheet for 'delete account' confirmation instead of dialog.

Note that I have also removed all of the items from the main and edit menus that have moved into the preferences pain (except font size per our previous discussion) I was gonna do that in another PR but given the amount of time to update on this figured getting it all in one PR was  better.

One question for you sly.  Currently when you click on the Delete Account button the preferences window closes and then a confirmation alert appears.  @jleandroperez had thought this  felt a little weird and wondered if the preferences pain should stay open.  Thoughts?

@jleandroperez @pachlava  this is ready for a code review as well when you have a moment.

**Comment by SylvesterWilmott** [comment]
Looks good! Thanks @charliescheer 

It looks like there's a tiny error in the button label, currently it's sentence case `Log out`. Let's update this to title case `Log Out`.

>One question for you sly. Currently when you click on the Delete Account button the preferences window closes and then a confirmation alert appears. @jleandroperez had thought this felt a little weird and wondered if the preferences pain should stay open. Thoughts?

I see the checked item `Use a sheet for 'delete account' confirmation instead of dialog.` but maybe I'm missing something? I figured this was the purpose of using a sheet attached to the preferences to confirm account deletion rather than the dialog? I'm going off the build you sent me but the dialog we display after pressing "Delete Account" should be shown attached to the preferences and not the main app window.

**Comment by pachlava** [comment]
Thanks for addressing all the issues @charliescheer! I spent around 30 minutes testing, doing various kinds of stuff I could come up with, like having Preferences opened in the background, while using shortcuts to adjust font size, to make sure the slider position is updated. Also, tried testing in full screen and under different themes, as well as running all checks that you provided in the PR description. Found no issues 👍

**Comment by charliescheer** [comment]
>I see the checked item Use a sheet for 'delete account' confirmation instead of dialog. but maybe I'm missing something? I figured this was the purpose of using a sheet attached to the preferences to confirm account deletion rather than the dialog? I'm going off the build you sent me but the dialog we display after pressing "Delete Account" should be shown attached to the preferences and not the main app window.

Gotcha.  Thanks for the additional details.  I misunderstood what everyone was thinking with that note.  I checked that the warning appeared as a sheet, not what it was attached to. I've made that change and will send you a new version to test.  Thanks for checking it out!

**Comment by SylvesterWilmott** [comment]
Thanks for the updates. Design look good to me! Thanks @charliescheer

**Review by jleandroperez — APPROVED** [review]
@charliescheer I found a nipticky detail. 

Clicking on the Analytics link, at the bottom of the UI, causes the text to shrink. Pasting a clip below (not sure why is that?)

Other than that, great job!!

:shipit: when ready



https://user-images.githubusercontent.com/1195260/139849585-8072194d-0919-4d6a-a5b8-ea13f2ebb68a.mov

**Comment by charliescheer** [comment]
Thanks for taking a look @jleandroperez 

>Clicking on the Analytics link, at the bottom of the UI, causes the text to shrink. Pasting a clip below (not sure why is that?)

Turns out it does that because the NSTextField and the NSMutableAttributedString are [trying to set a font to that text.](https://stackoverflow.com/questions/24318032/nsattributedstring-in-textfield-changing-resetting-on-click)  When it is clicked on the attributed string's font takes over and the default font for the attributed string is not the system font.  

Fixed it by making the attributed string font explicit.  Gonna merge this change.

**Pull request merged**